feat(concerto-cto): Migrate package to TypeScript

imfeniljikadara opened the pull request:
# feat(concerto-cto): Migrate package to TypeScript

# Closes #406 

This PR implements the TypeScript migration for the concerto-cto package, converting JavaScript files to TypeScript while maintaining all existing functionality and test coverage.

### Changes
- Converted core JavaScript files to TypeScript:
  - `parseexception.js` → `parseexception.ts`
  - `parserMain.js` → `parserMain.ts`
  - `printer.js` → `printer.ts`
  - `external.js` → `external.ts`
  - `index.js` → `index.ts`
- Added comprehensive type definitions and interfaces
- Configured TypeScript build settings with strict type checking
- Maintained CommonJS module compatibility
- Added type declarations for external dependencies
- Preserved all existing functionality and API contracts

### Flags
- Code coverage metrics are currently below threshold due to TypeScript-specific constructs
- Need assistance with tuning coverage exclusion rules
- All 136 existing tests are passing
- Build process successfully compiles TypeScript to JavaScript

### Screenshots or Video
<img width="695" alt="Screenshot 2025-03-27 at 10 06 59 AM" src="https://github.com/user-attachments/assets/b90e6e6e-c20d-4943-a6f7-ff2b5ac97bcc" />

This coverage drop is expected during TypeScript migration due to added interfaces and type definitions.

### Related Issues
- Issue #405
- Pull Request 

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [x] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [x] Extend the documentation, if necessary
- [x] Merging to `main` from `fork:branchname`

mttrbrts commented on packages/concerto-cto/src/external.ts:
```diff
@@ -0,0 +1,100 @@
+
+const debugLog = debug('concerto:ModelManager');
+
+interface Model {
```
These interfaces are already defined in the concerto-types package. Please import those instead.

mttrbrts replied:
Can these interfaces be removed now?

mttrbrts reviewed (commented):
Let's also delete the old JS files, that will help the code coverage

imfeniljikadara commented:
Hi @mttrbrts , I've made two improvements to help with the coverage issue:

1. Deleted the old JavaScript files from `concerto-cto`.
2. Updated `external.ts` to use the proper interfaces from `concerto-types` package.  replaced local `Model` and `Models` interfaces with `IModel` and `IModels` from `@accordproject/concerto-types`

imfeniljikadara commented on packages/concerto-cto/src/parseexception.ts:
```diff
@@ -28,12 +26,12 @@ class ParseException extends BaseFileException {
      * @param {string} [component] - the component which throws this error
      */
-    constructor(message, fileLocation, fileName, fullMessageOverride, component) {
+    constructor(message: string, fileLocation?: any, fileName?: string, fullMessageOverride?: string, component?: string) {
```
I've tightened the typing for the `fileLocation` parameter by replacing `any` with a specific interface that matches how it's used in the code.

imfeniljikadara commented on packages/concerto-cto/src/parserMain.ts:
```diff
@@ -47,8 +49,8 @@ function parse(cto, fileName, options) {
  */
-function parseModels(files, options) {
-    const result = {
+function parseModels(files: string[], options?: ParseOptions): any {
```
I've updated the return type of the `parseModels` function from `any` to `IModels` and added the necessary import from '`@accordproject/concerto-types`'.

imfeniljikadara commented on packages/concerto-cto/src/printer.ts:
```diff
@@ -12,25 +12,109 @@
+
+interface MapEntry {
+    $class: string;
+    key: any;
+    value: any;
+    decorators?: Decorator[];
+}
+
+interface Declaration {
+    $class: string;
+    name?: string;
+    isAbstract?: boolean;
+    superType?: { name: string };
+    properties?: Property[];
+    decorators?: Decorator[];
+    values?: any[];
+    range?: { lower: number; upper: number };
+    keyType?: any;
+    valueType?: any;
+    isIdentified?: boolean;
+    identifierFieldName?: string;
+}
+
+interface Import {
+    $class: string;
+    namespace: string;
+    uri?: string;
+    name?: string;
+    types?: string[];
+    aliasedTypes?: Array<{ name: string; aliasedName: string }>;
+}
+
+interface MetaModel {
+    $class: string;
+    namespace: string;
+    imports?: Import[];
+    declarations?: Declaration[];
+    concertoVersion?: string;
+    decorators?: Decorator[];
+}
```
I've updated the `printer.ts` file to use interfaces from the `concerto-types` package instead of defining them locally.

imfeniljikadara commented on packages/concerto-cto/src/printer.ts:
```diff
@@ -12,25 +12,109 @@
  */
-function isMap(mm) {
-    return  mm.$class === `${MetaModelNamespace}.MapDeclaration`;
+function isMap(mm: any): boolean {
```
I've updated the `isMap` function to use `IDeclaration` as the parameter type instead of `IModel`. 

**Why I have used `IDeclaration` type instead of `IModel`:**

The problem with using `IModel` for the `isMap` functions is that an `IModel` and an `IDeclaration` represent different concepts:
- `IModel` represents a complete model with namespace, imports, and declarations
- `IDeclaration` represents a single declaration within a model (like a class, map, or enum declaration)
- `IDeclaration` is the correct interface for the objects being passed to the function and function is checking for specific declaration type (MapDeclaration). an `IModel` actually contains declarations but isn't itself a declaration

If we want to use `IModel`, we would need to change these functions to check if the model contains a certain type of declaration, rather than if a declaration is of a certain type.

imfeniljikadara commented on packages/concerto-cto/src/printer.ts:
```diff
@@ -12,25 +12,109 @@
  * @return {boolean} the string for that model
  */
-function isScalar(mm) {
+function isScalar(mm: any): boolean {
```
I've updated the `isScalar` function to use `IDeclaration` as the parameter type instead of `IModel`. 

**Why I have used `IDeclaration` type instead of `IModel`:**

The problem with using `IModel` for the `isScalar` functions is that an `IModel` and an `IDeclaration` represent different concepts:
- `IModel` represents a complete model with namespace, imports, and declarations
- `IDeclaration` represents a single declaration within a model (like a class, map, or enum declaration)
- `IDeclaration` is the correct interface for the objects being passed to the function and function is checking for specific declaration type (ScalarDeclaration). an `IModel` actually contains declarations but isn't itself a declaration

If we want to use `IModel`, we would need to change these functions to check if the model contains a certain type of declaration, rather than if a declaration is of a certain type.

imfeniljikadara commented on packages/concerto-cto/src/printer.ts:
```diff
@@ -232,47 +321,47 @@ function modifiersFromMetaModel(mm){
         result += 'o';
     }
-    result += typeFromMetaModel(prop);
+    result += typeFromMetaModel(prop as any as MetaModelWithType);
```
I've changed the `typeFromMetaModel` function to accept `any` as its parameter type, now we can directly pass `prop` to the function without any casting.

imfeniljikadara commented on packages/concerto-cto/src/printer.ts:
```diff
@@ -232,47 +321,47 @@ function modifiersFromMetaModel(mm){
         result += decoratorsFromMetaModel(prop.decorators, '  ');
     }
-    if (prop.$class === `${MetaModelNamespace}.RelationshipProperty`) {
+    if (prop.isRelationship) {
```
I've simplified the code to directly check the `$class` property, which is the correct approach.

imfeniljikadara commented on packages/concerto-cto/src/types/concerto-metamodel.d.ts:
I've remove the custom type declarations. The imports are directly from `@accordproject/concerto-metamodel` as they should be.

imfeniljikadara commented on packages/concerto-cto/package.json:
```diff
@@ -55,19 +57,20 @@
     "ts-node": "10.9.1",
-    "typescript": "4.6.3"
+    "typescript": "4.6.3",
+    "webpack-cli": "^6.0.1"
```
`webpack-cli` and `path-browserify` are included because:
1. This package needs to run in both Node.js and browsers
2. `webpack-cli` builds the browser-compatible bundle
3. `path-browserify` provides browser implementations of Node's path module functions
4. The webpack config specifically requires `path-browserify` as a fallback since Webpack 5 no longer automatically includes Node.js polyfills

imfeniljikadara commented:
Hi @mttrbrts ,

Thank you for your valuable feedback and guidance! i've addressed all your comments, applied the necessary updates, and incorporated the suggested changes. now, all tests are passing successfully, and the coverage is in great shape.

Let me know if there's anything else you'd like me to refine or review further. looking forward to your feedback.

<img width="692" alt="Screenshot 2025-03-28 at 8 29 29 PM" src="https://github.com/user-attachments/assets/ef02cbfb-bf63-4f80-8d45-b8e8b8a2d656" />

mttrbrts commented on packages/concerto-cto/package.json:
```diff
@@ -55,19 +57,20 @@
     "@accordproject/concerto-metamodel": "3.8.1",
-    "@accordproject/concerto-util": "^4.0.0-alpha.0",
-    "path-browserify": "1.0.1"
+    "@accordproject/concerto-util": "^4.0.0-alpha.0"
```
v4 is an unstable release, please use the same v3.x version that's defined in this monorepo for the concerto-util package

imfeniljikadara replied:
I've updated `package.json` file to use the stable version `3.9.1` of `concerto-util`.

mttrbrts commented on packages/concerto-cto/src/parserMain.ts:
```diff
@@ -45,10 +48,10 @@ function parse(cto, fileName, options) {
-function parseModels(files, options) {
-    const result = {
+function parseModels(files: string[], options?: ParseOptions): IModels {
+    const result: { $class: string; models: any[] } = {
```
```suggestion
    const result: IModels = {
```

imfeniljikadara replied:
You're absolutely right! i must use the `IModels` type consistently throughout the code.

mttrbrts reviewed (changes requested):
Nearly there @imfeniljikadara! Thanks for making the fixes.

I've added a few more minor comments, please also look at the build, it looks like this has introduced some TypeScript compilation errors.
https://github.com/accordproject/concerto/actions/runs/14131673919/job/39639823055?pr=1005#step:5:66

Also, do you understand the reason for the dropped code-coverage? My hope is that we can match the existing coverage thresholds. Thanks

imfeniljikadara commented:
Hi @mttrbrts ,

Thank you for your feedback and guidance! i've addressed all the changes, resolved the compilation errors, and added two specific tests for `printer.ts`:
- Should handle ImportTypes with aliased types (Line 430-431 of `printer.ts`)
- Should handle ImportAll type (Line 436 of `printer.ts`)

All tests are now passing (138 passing), and coverage is nearly perfect:

<img width="688" alt="Screenshot 2025-03-30 at 8 22 13 PM" src="https://github.com/user-attachments/assets/eb668130-297b-47b7-9cc9-eb0d890b6ef1" />

The uncovered cases in `printer.ts` (Lines: 66, 155-161, 239, 329, 384, 442 of `printer.ts`) seem to be impacting the branch coverage. these seem related to specific conditions that might not be covered by existing tests. 

<img width="986" alt="Screenshot 2025-03-30 at 8 23 30 PM" src="https://github.com/user-attachments/assets/8d89794a-e287-4623-b95b-1175fa52c22e" />

I would really appreciate any insights on how best to refine these cases to meet the required threshold.

imfeniljikadara commented:
Hi @mttrbrts ,

I’ve resolved the `npm ci` EUSAGE error and ensured dependency consistency. all tests are still passing, and coverage remains the same. Let me know if there's anything else to address. Thank you!

mttrbrts commented on packages/concerto-cto/src/printer.ts:
```diff
@@ -137,11 +152,15 @@ function typeFromMetaModel(mm){
     case `${MetaModelNamespace}.ObjectMapKeyType`:
     case `${MetaModelNamespace}.ObjectMapValueType`:
-        result += ` ${mm.type.name}`;
+        if (mm.type) {
```
Did the compiler insist on this?

All of the three nodes in this case have a non-optional `type` property, so the condition shouldn't be necessary.

I don't understand why the coverage shows a miss because we have a test model that corresponds to this case. Perhaps it's because we don't test the negative case? Although that's not possible, because the property is mandatory. 😕 

I'm not able to pull your code to test this out locally, can you try this, please? https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

imfeniljikadara replied:
Yeah, the compiler did insist on that line ☹️, which is why I added the check. i also found the coverage miss confusing since we already have a test model for this case. your theory about the negative case makes sense, but given that the property is mandatory, it shouldn't be possible. 🤔

Sorry Matt, I realized my PR was still in draft mode, which is why you weren’t able to pull the code. i've now marked it as **Ready for Review**, so you should be able to make edits. Let me know if it works!

mttrbrts commented on packages/concerto-cto/package.json:
```diff
@@ -55,19 +57,21 @@
     "@accordproject/concerto-metamodel": "3.8.1",
-    "@accordproject/concerto-util": "^4.0.0-alpha.0",
-    "path-browserify": "1.0.1"
+    "@accordproject/concerto-types": "^3.8.1",
+    "@accordproject/concerto-util": "3.9.1"
```
Sorry @imfeniljikadara, I was wrong here before. I forgot that you're targeting the v4 branch. These dependencies should point to the v4.0.0-alpha.0 release!

imfeniljikadara replied:
Thanks for the clarification, and no worries Matt, you are absolutely right that it should be pointing out `v4.0.0-alpha.0`

mttrbrts reviewed (commented):
Thanks for the quick turn-around @imfeniljikadara! 

I want to debug the coverage issue, because that implies something wrong that the compiler is picking up. The existing tests should cover (most of) the missing lines.

There are a few `any` types, however, I'm happy to commit those and tighten as we integrate the rest of the packages.

imfeniljikadara commented:
Hi @mttrbrts ,

Thank you for the quick response, appreciate your guidance! that makes sense. debugging the coverage issue should help us understand what the compiler is picking up. since the existing tests should already cover most cases, i’ll take another look as well to see if anything stands out.

Also Matt, i’ve marked the PR as **Ready for Review**, so you should now be able to pull the code and make any necessary changes. if there's anything specific you'd like me to revisit or refine, just let me know.

mttrbrts reviewed (approved):
Thanks again @imfeniljikadara, I was able to pull the branch, and make a few minor changes. We're now at 100% branch coverage too, without changing any tests.

Good work, and thanks for your contribution.
